File: stitch_agent/run/repo_context.py
"""Analyze a repository to detect language, framework, and CI configuration."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pathlib import Path


@dataclass
class RepoContext:
    """Detected repository context for CI generation."""

    languages: list[str] = field(default_factory=list)
    package_manager: str | None = None
    frameworks: list[str] = field(default_factory=list)
    ci_platform: str | None = None
    has_test_jobs: bool = False
    existing_ci_file: str | None = None
    entry_files: list[str] = field(default_factory=list)

    def summary(self) -> str:
        parts = []
        if self.languages:
            parts.append(f"Languages: {', '.join(self.languages)}")
        if self.package_manager:
            parts.append(f"Package manager: {self.package_manager}")
        if self.frameworks:
            parts.append(f"Frameworks/tools: {', '.join(self.frameworks)}")
        if self.ci_platform:
            parts.append(f"CI platform: {self.ci_platform}")
        if self.existing_ci_file:
            parts.append(f"Existing CI config: {self.existing_ci_file}")
        parts.append(f"Has test jobs: {'yes' if self.has_test_jobs else 'no'}")
        if self.entry_files:
            parts.append(f"Key files: {', '.join(self.entry_files[:10])}")
        return "\n".join(parts)


# Language detection: config file -> (language, package_manager, frameworks)
_LANG_SIGNALS: list[tuple[str, str, str | None, list[str]]] = [
    ("pyproject.toml", "python", "uv", []),
    ("setup.py", "python", "pip", []),
    ("setup.cfg", "python", "pip", []),
    ("requirements.txt", "python", "pip", []),
    ("Pipfile", "python", "pipenv", []),
    ("package.json", "javascript", None, []),  # pm detected separately
    ("bun.lockb", "javascript", "bun", []),
    ("pnpm-lock.yaml", "javascript", "pnpm", []),
    ("yarn.lock", "javascript", "yarn", []),
    ("package-lock.json", "javascript", "npm", []),
    ("go.mod", "go", "go", []),
    ("Cargo.toml", "rust", "cargo", []),
    ("Gemfile", "ruby", "bundler", []),
    ("composer.json", "php", "composer", []),
]

# Framework detection from config files
_FRAMEWORK_SIGNALS: list[tuple[str, str]] = [
    ("pytest.ini", "pytest"),
    ("conftest.py", "pytest"),
    ("jest.config.js", "jest"),
    ("jest.config.ts", "jest"),
    ("vitest.config.ts", "vitest"),
    ("vitest.config.js", "vitest"),
    (".eslintrc", "eslint"),
    (".eslintrc.json", "eslint"),
    ("eslint.config.js", "eslint"),
    ("biome.json", "biome"),
    ("biome.jsonc", "biome"),
    (".prettierrc", "prettier"),
    ("tsconfig.json", "typescript"),
    ("ruff.toml", "ruff"),
    ("mypy.ini", "mypy"),
    (".golangci.yml", "golangci-lint"),
    (".golangci.yaml", "golangci-lint"),
]

# CI platforms
_CI_CONFIGS: list[tuple[str, str]] = [
    (".gitlab-ci.yml", "gitlab"),
    (".github/workflows", "github"),
    ("bitbucket-pipelines.yml", "bitbucket"),
    (".circleci/config.yml", "circleci"),
    (".travis.yml", "travis"),
    ("Jenkinsfile", "jenkins"),
    ("azure-pipelines.yml", "azure"),
]

# Test job name patterns (to detect if test jobs already exist)
_TEST_JOB_PATTERNS = {"test", "lint", "check", "typecheck", "audit", "format"}


def analyze_repo(repo_root: Path) -> RepoContext:
    """Analyze a repository and return detected context."""
    ctx = RepoContext()

    # Detect languages and package manager
    seen_langs: set[str] = set()
    for filename, lang, pm, frameworks in _LANG_SIGNALS:
        if (repo_root / filename).exists():
            if lang not in seen_langs:
                ctx.languages.append(lang)
                seen_langs.add(lang)
            if pm and not ctx.package_manager:
                ctx.package_manager = pm
            ctx.frameworks.extend(f for f in frameworks if f not in ctx.frameworks)

    # Detect frameworks
    for filename, framework in _FRAMEWORK_SIGNALS:
        if (repo_root / filename).exists() and framework not in ctx.frameworks:
            ctx.frameworks.append(framework)

    # Detect from pyproject.toml dependencies
    pyproject = repo_root / "pyproject.toml"
    if pyproject.is_file():
        _detect_from_pyproject(pyproject, ctx)

    # Detect from package.json
    pkg_json = repo_root / "package.json"
    if pkg_json.is_file():
        _detect_from_package_json(pkg_json, ctx)

    # Detect CI platform
    for path, platform in _CI_CONFIGS:
        full = repo_root / path
        if full.exists():
            ctx.ci_platform = platform
            if full.is_file():
                ctx.existing_ci_file = path
            elif full.is_dir():
                # For GitHub workflows, find first yml
                ymls = sorted(full.glob("*.yml")) + sorted(full.glob("*.yaml"))
                if ymls:
                    ctx.existing_ci_file = str(ymls[0].relative_to(repo_root))
            break

    # Detect if test jobs already exist
    if ctx.existing_ci_file and ctx.ci_platform:
        ctx.has_test_jobs = _has_test_jobs(repo_root, ctx.ci_platform)

    # Collect key config files
    for name in [
        "pyproject.toml", "package.json", "go.mod", "Cargo.toml",
        "Makefile", "Dockerfile", "docker-compose.yml",
        "tsconfig.json", "biome.json",
    ]:
        if (repo_root / name).is_file():
            ctx.entry_files.append(name)

    return ctx


def _detect_from_pyproject(path: Path, ctx: RepoContext) -> None:
    """Extract framework info from pyproject.toml (simple text scanning)."""
    try:
        text = path.read_text()
    except OSError:
        return

    if "pytest" in text and "pytest" not in ctx.frameworks:
        ctx.frameworks.append("pytest")
    if "ruff" in text and "ruff" not in ctx.frameworks:
        ctx.frameworks.append("ruff")
    if "mypy" in text and "mypy" not in ctx.frameworks:
        ctx.frameworks.append("mypy")
    if "hatchling" in text or "hatch" in text:
        ctx.package_manager = ctx.package_manager or "hatch"


def _detect_from_package_json(path: Path, ctx: RepoContext) -> None:
    """Extract framework info from package.json."""
    import json

    try:
        data = json.loads(path.read_text())
    except (OSError, json.JSONDecodeError):
        return

    all_deps = {}
    for key in ("dependencies", "devDependencies"):
        all_deps.update(data.get(key, {}))

    dep_framework_map = {
        "jest": "jest",
        "vitest": "vitest",
        "mocha": "mocha",
        "eslint": "eslint",
        "@biomejs/biome": "biome",
        "prettier": "prettier",
        "typescript": "typescript",
    }
    for dep, framework in dep_framework_map.items():
        if dep in all_deps and framework not in ctx.frameworks:
            ctx.frameworks.append(framework)

    # Detect scripts
    scripts = data.get("scripts", {})
    if "test" in scripts and "test" not in ctx.frameworks:
        ctx.entry_files.append(f"package.json scripts.test: {scripts['test']}")
    if "lint" in scripts:
        ctx.entry_files.append(f"package.json scripts.lint: {scripts['lint']}")


def _has_test_jobs(repo_root: Path, platform: str) -> bool:
    """Check if the CI config already defines test/lint jobs."""
    import yaml

    if platform == "gitlab":
        ci_file = repo_root / ".gitlab-ci.yml"
        if not ci_file.is_file():
            return False
        try:
            data = yaml.safe_load(ci_file.read_text())
        except yaml.YAMLError:
            return False
        if not isinstance(data, dict):
            return False
        for key in data:
            if isinstance(key, str) and any(p in key.lower() for p in _TEST_JOB_PATTERNS):
                return True
        return False

    if platform == "github":
        wf_dir = repo_root / ".github" / "workflows"
        if not wf_dir.is_dir():
            return False
        for wf in wf_dir.glob("*.yml"):
            try:
                data = yaml.safe_load(wf.read_text())
            except yaml.YAMLError:
                continue
            if not isinstance(data, dict):
                continue
            jobs = data.get("jobs", {})
            if isinstance(jobs, dict):
                for job_name in jobs:
                    if any(p in job_name.lower() for p in _TEST_JOB_PATTERNS):
                        return True
        return False

    return False

File: src/core/repo-context.ts
import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { join } from "node:path";

interface RepoContext {
  languages: string[];
  packageManager: string | null;
  frameworks: string[];
  ciPlatform: string | null;
  hasTestJobs: boolean;
  existingCIFile: string | null;
  entryFiles: string[];
}

const LANG_SIGNALS: [string, string, string | null, string[]][] = [
  ["pyproject.toml", "python", "uv", []],
  ["setup.py", "python", "pip", []],
  ["setup.cfg", "python", "pip", []],
  ["requirements.txt", "python", "pip", []],
  ["Pipfile", "python", "pipenv", []],
  ["package.json", "javascript", null, []],
  ["bun.lockb", "javascript", "bun", []],
  ["bun.lock", "javascript", "bun", []],
  ["pnpm-lock.yaml", "javascript", "pnpm", []],
  ["yarn.lock", "javascript", "yarn", []],
  ["package-lock.json", "javascript", "npm", []],
  ["go.mod", "go", "go", []],
  ["Cargo.toml", "rust", "cargo", []],
  ["Gemfile", "ruby", "bundler", []],
  ["composer.json", "php", "composer", []],
];

const FRAMEWORK_SIGNALS: [string, string][] = [
  ["pytest.ini", "pytest"],
  ["conftest.py", "pytest"],
  ["jest.config.js", "jest"],
  ["jest.config.ts", "jest"],
  ["vitest.config.ts", "vitest"],
  ["vitest.config.js", "vitest"],
  [".eslintrc", "eslint"],
  [".eslintrc.json", "eslint"],
  ["eslint.config.js", "eslint"],
  ["biome.json", "biome"],
  ["biome.jsonc", "biome"],
  [".prettierrc", "prettier"],
  ["tsconfig.json", "typescript"],
  ["ruff.toml", "ruff"],
  ["mypy.ini", "mypy"],
  [".golangci.yml", "golangci-lint"],
  [".golangci.yaml", "golangci-lint"],
];

const CI_CONFIGS: [string, string][] = [
  [".gitlab-ci.yml", "gitlab"],
  [".github/workflows", "github"],
  ["bitbucket-pipelines.yml", "bitbucket"],
  [".circleci/config.yml", "circleci"],
  [".travis.yml", "travis"],
  ["Jenkinsfile", "jenkins"],
  ["azure-pipelines.yml", "azure"],
];

const TEST_JOB_PATTERNS = new Set(["test", "lint", "check", "typecheck", "audit", "format"]);

function detectLanguages(repoRoot: string, ctx: RepoContext): void {
  const seenLangs = new Set<string>();

  // Detect languages and package manager
  for (const [filename, lang, pm] of LANG_SIGNALS) {
    if (existsSync(join(repoRoot, filename))) {
      if (!seenLangs.has(lang)) {
        ctx.languages.push(lang);
        seenLangs.add(lang);
      }
      if (pm && !ctx.packageManager) {
        ctx.packageManager = pm;
      }
    }
  }
}

function detectFrameworks(repoRoot: string, ctx: RepoContext): void {
  // Detect frameworks
  for (const [filename, framework] of FRAMEWORK_SIGNALS) {
    if (existsSync(join(repoRoot, filename)) && !ctx.frameworks.includes(framework)) {
      ctx.frameworks.push(framework);
    }
  }

  // Detect from pyproject.toml
  const pyproject = join(repoRoot, "pyproject.toml");
  if (existsSync(pyproject)) {
    detectFromPyproject(pyproject, ctx);
  }

  // Detect from package.json
  const pkgJson = join(repoRoot, "package.json");
  if (existsSync(pkgJson)) {
    detectFromPackageJson(pkgJson, ctx);
  }
}

function resolveExistingCIFile(full: string, relPath: string): string | null {
  try {
    if (statSync(full).isFile()) return relPath;
    if (statSync(full).isDirectory()) {
      const ymls = readdirSync(full)
        .filter((f) => f.endsWith(".yml") || f.endsWith(".yaml"))
        .sort();
      if (ymls.length > 0) return `${relPath}/${ymls[0]}`;
    }
  } catch {
    // ignore
  }
  return null;
}

function detectCIPlatform(repoRoot: string, ctx: RepoContext): void {
  for (const [path, platform] of CI_CONFIGS) {
    const full = join(repoRoot, path);
    if (!existsSync(full)) continue;
    ctx.ciPlatform = platform;
    ctx.existingCIFile = resolveExistingCIFile(full, path);
    break;
  }
}

function collectEntryFiles(repoRoot: string, ctx: RepoContext): void {
  // Collect key config files
  for (const name of [
    "pyproject.toml",
    "package.json",
    "go.mod",
    "Cargo.toml",
    "Makefile",
    "Dockerfile",
    "docker-compose.yml",
    "tsconfig.json",
    "biome.json",
  ]) {
    if (existsSync(join(repoRoot, name))) {
      ctx.entryFiles.push(name);
    }
  }
}

export function analyzeRepo(repoRoot: string): RepoContext {
  const ctx: RepoContext = {
    languages: [],
    packageManager: null,
    frameworks: [],
    ciPlatform: null,
    hasTestJobs: false,
    existingCIFile: null,
    entryFiles: [],
  };

  detectLanguages(repoRoot, ctx);
  detectFrameworks(repoRoot, ctx);
  detectCIPlatform(repoRoot, ctx);

  // Detect test jobs
  if (ctx.existingCIFile && ctx.ciPlatform) {
    ctx.hasTestJobs = hasTestJobs(repoRoot, ctx.ciPlatform);
  }

  collectEntryFiles(repoRoot, ctx);

  return ctx;
}

export function repoContextSummary(ctx: RepoContext): string {
  const parts: string[] = [];
  if (ctx.languages.length > 0) parts.push(`Languages: ${ctx.languages.join(", ")}`);
  if (ctx.packageManager) parts.push(`Package manager: ${ctx.packageManager}`);
  if (ctx.frameworks.length > 0) parts.push(`Frameworks/tools: ${ctx.frameworks.join(", ")}`);
  if (ctx.ciPlatform) parts.push(`CI platform: ${ctx.ciPlatform}`);
  if (ctx.existingCIFile) parts.push(`Existing CI config: ${ctx.existingCIFile}`);
  parts.push(`Has test jobs: ${ctx.hasTestJobs ? "yes" : "no"}`);
  if (ctx.entryFiles.length > 0) parts.push(`Key files: ${ctx.entryFiles.slice(0, 10).join(", ")}`);
  return parts.join("\n");
}

function detectFromPyproject(path: string, ctx: RepoContext): void {
  let text: string;
  try {
    text = readFileSync(path, "utf-8");
  } catch {
    return;
  }
  if (text.includes("pytest") && !ctx.frameworks.includes("pytest")) ctx.frameworks.push("pytest");
  if (text.includes("ruff") && !ctx.frameworks.includes("ruff")) ctx.frameworks.push("ruff");
  if (text.includes("mypy") && !ctx.frameworks.includes("mypy")) ctx.frameworks.push("mypy");
  if ((text.includes("hatchling") || text.includes("hatch")) && !ctx.packageManager) {
    ctx.packageManager = "hatch";
  }
}

function detectFromPackageJson(path: string, ctx: RepoContext): void {
  let data: Record<string, unknown>;
  try {
    data = JSON.parse(readFileSync(path, "utf-8"));
  } catch {
    return;
  }

  const allDeps: Record<string, unknown> = {
    ...(typeof data.dependencies === "object" && data.dependencies !== null
      ? (data.dependencies as Record<string, unknown>)
      : {}),
    ...(typeof data.devDependencies === "object" && data.devDependencies !== null
      ? (data.devDependencies as Record<string, unknown>)
      : {}),
  };

  const depMap: Record<string, string> = {
    jest: "jest",
    vitest: "vitest",
    mocha: "mocha",
    eslint: "eslint",
    "@biomejs/biome": "biome",
    prettier: "prettier",
    typescript: "typescript",
  };

  for (const [dep, framework] of Object.entries(depMap)) {
    if (dep in allDeps && !ctx.frameworks.includes(framework)) {
      ctx.frameworks.push(framework);
    }
  }

  const scripts =
    typeof data.scripts === "object" && data.scripts !== null
      ? (data.scripts as Record<string, unknown>)
      : {};
  if ("test" in scripts) {
    ctx.entryFiles.push(`package.json scripts.test: ${scripts.test}`);
  }
  if ("lint" in scripts) {
    ctx.entryFiles.push(`package.json scripts.lint: ${scripts.lint}`);
  }
}

function hasTestJobsGitlab(repoRoot: string): boolean {
  const ciFile = join(repoRoot, ".gitlab-ci.yml");
  if (!existsSync(ciFile)) return false;
  try {
    const yaml = require("js-yaml");
    const data = yaml.load(readFileSync(ciFile, "utf-8"));
    if (typeof data !== "object" || data === null) return false;
    for (const key of Object.keys(data as Record<string, unknown>)) {
      if (
        typeof key === "string" &&
        [...TEST_JOB_PATTERNS].some((p) => key.toLowerCase().includes(p))
      ) {
        return true;
      }
    }
  } catch {
    return false;
  }
  return false;
}

function matchesTestPattern(name: string): boolean {
  const lower = name.toLowerCase();
  return [...TEST_JOB_PATTERNS].some((p) => lower.includes(p));
}

function hasTestJobsGithub(repoRoot: string): boolean {
  const wfDir = join(repoRoot, ".github", "workflows");
  if (!existsSync(wfDir)) return false;
  try {
    const yaml = require("js-yaml");
    for (const file of readdirSync(wfDir).filter((f) => f.endsWith(".yml"))) {
      const data = yaml.load(readFileSync(join(wfDir, file), "utf-8"));
      if (typeof data !== "object" || data === null) continue;
      const jobs = (data as Record<string, unknown>).jobs;
      if (typeof jobs !== "object" || jobs === null) continue;
      if (Object.keys(jobs as Record<string, unknown>).some(matchesTestPattern)) return true;
    }
  } catch {
    return false;
  }
  return false;
}

function hasTestJobs(repoRoot: string, platform: string): boolean {
  if (platform === "gitlab") return hasTestJobsGitlab(repoRoot);
  if (platform === "github") return hasTestJobsGithub(repoRoot);
  return false;
}
